scipy/stats/stats.py

```diff
@@ -5378,48 +5412,48 @@ def kstest(rvs, cdf, args=(), N=20, alternative='two-sided', mode='approx'):
     (0.131016895759829, 0.058826222555312224)
 
     """
+    if alternative not in ['two-sided', 'greater', 'less', 'two_sided']:
+        raise ValueError("Unexpected alternative %s" % alternative)
+    # to not break compatibility with existing code
+    if alternative == 'two_sided':
+        alternative = 'two-sided'
     if isinstance(rvs, str):
         if (not cdf) or (cdf == rvs):
             cdf = getattr(distributions, rvs).cdf
             rvs = getattr(distributions, rvs).rvs
         else:
             raise AttributeError("if rvs is string, cdf has to be the "
                                  "same distribution")
-
     if isinstance(cdf, str):
         cdf = getattr(distributions, cdf).cdf
-    if callable(rvs):
-        kwds = {'size': N}
-        vals = np.sort(rvs(*args, **kwds))
-    else:
-        vals = np.sort(rvs)
-        N = len(vals)
+
+    vals = np.sort(rvs(*args, size=N) if callable(rvs) else rvs)
+    N = len(vals)
     cdfvals = cdf(vals, *args)
 
-    # to not break compatibility with existing code
-    if alternative == 'two_sided':
-        alternative = 'two-sided'
+    if alternative == 'greater':
+        Dplus = _compute_dplus(cdfvals)
+        return KstestResult(Dplus, distributions.ksone.sf(Dplus, N))
 
-    if alternative in ['two-sided', 'greater']:
-        Dplus = (np.arange(1.0, N + 1)/N - cdfvals).max()
-        if alternative == 'greater':
-            return KstestResult(Dplus, distributions.ksone.sf(Dplus, N))
-
-    if alternative in ['two-sided', 'less']:
-        Dmin = (cdfvals - np.arange(0.0, N)/N).max()
-        if alternative == 'less':
-            return KstestResult(Dmin, distributions.ksone.sf(Dmin, N))
-
-    if alternative == 'two-sided':
-        D = np.max([Dplus, Dmin])
-        if mode == 'asymp':
-            return KstestResult(D, distributions.kstwobign.sf(D * np.sqrt(N)))
-        if mode == 'approx':
-            pval_two = distributions.kstwobign.sf(D * np.sqrt(N))
-            if N > 2666 or pval_two > 0.80 - N*0.3/1000:
-                return KstestResult(D, pval_two)
-            else:
-                return KstestResult(D, 2 * distributions.ksone.sf(D, N))
+    if alternative == 'less':
+        Dminus = _compute_dminus(cdfvals)
+        return KstestResult(Dminus, distributions.ksone.sf(Dminus, N))
+
+    # alternative == 'two-sided':
+    Dplus = _compute_dplus(cdfvals)
+    Dminus = _compute_dminus(cdfvals)
+    D = np.max([Dplus, Dminus])
+    if mode == 'auto':
+        mode = 'exact' if N < 1000000 else 'asymp'
```

Comment:
1e6 seems to be very large, wouldn't the asymptotic formula work well for much smaller N, e.g. N=1000?

Reply:
Upon reflection, it's not clear to me whether this switchover needs to exist any more.  The `exact` path calls `kstwo.sf(D, N)` which knows when to approximate with the asymptotic.  The asymptotic likely matches well for large `N`, except for small `D`, those in the range `[0, 2/N]` say, as the exact value approaches `1` faster than the asymptotic.

Reply:
agree, one could even consider to use 'exact' as the default option?

Reply:
the current version is:
```if mode == 'auto':  # Always select exact
    mode = 'exact'
```
i think auto is not even needed for backwards compatibility since it is introduced in this PR. should we go with exact as the default option or do I miss something?